### DISCUSSION TITLE
The configuration constraints.php does not work (anymore)

### DISCUSSION THREAD

**Discussion by back-2-95** [renovatebot/renovate#21204] [Suggest an Idea]
### Type of discussion.

I just want to chat

### Tell us more.

In our [Renovate testing repo](https://github.com/druidfi/renovate-test) we have set in config the following:

```json
{
  "constraints": {
    "php": "8.1.14"
  }
}
```

It worked at the time when that repo was created (January 2023) but not anymore.

Now PRs get this error:

```
Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v "/mnt/renovate/gh/druidfi/renovate-test":"/mnt/renovate/gh/druidfi/renovate-test" -v "/tmp/renovate-cache":"/tmp/renovate-cache" -v "/tmp/containerbase":"/tmp/containerbase" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w "/mnt/renovate/gh/druidfi/renovate-test" docker.io/containerbase/sidecar bash -l -c "install-tool php 8.2.4 && install-tool composer 2.5.5 && composer update drupal/core-composer-scaffold drupal/core-dev-pinned drupal/core-recommended --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins"
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - laminas/laminas-diactoros[2.14.0, ..., 2.14.x-dev] require php ^7.3 || ~8.0.0 || ~8.1.0 -> your php version (8.2.4) does not satisfy that requirement.
    - drupal/core-recommended 9.5.7 requires laminas/laminas-diactoros ~2.14.0 -> satisfiable by laminas/laminas-diactoros[2.14.0, 2.14.x-dev].
    - Root composer.json requires drupal/core-recommended 9.5.7 -> satisfiable by drupal/core-recommended[9.5.7].

Use the option --with-all-dependencies (-W) to allow upgrades, downgrades and removals for packages currently locked to specific versions.
```

and same from Renovate logs:

```
Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/druidfi/renovate-test\":\"/mnt/renovate/gh/druidfi/renovate-test\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/druidfi/renovate-test\" ghcr.io/containerbase/sidecar bash -l -c \"install-tool php 8.2.4 && install-tool composer 2.5.5 && composer update .....
```

- There is no `platform.php` in composer.json (which imo is used or understood wrongly by many)
- `require.php` is `^8.1`

So something changed for handling `constraints.php` in Renovate config.

I wanna achieve situation where Renovate install-tool installs PHP 8.1 for this project as.

**Comment by back-2-95** [comment]
And then I find this 

![image](https://user-images.githubusercontent.com/1140272/228125662-4868ba83-d3ad-4173-99c9-250f790daa4c.png)

**Comment by back-2-95** [comment]
And I now tested it with a few projects and I can say this works:

```json
{
  "force": {
    "constraints": {
      "php": "8.1.16"
    }
  }
}
```

**Reply by Wirone** [reply]
Thanks for this tip, but in my case it does not work even with `force`. I have:

- `"extractedConstraints": {"php": "8.2.*"}` from `composer.json`
- `"force": {"constraints": {"php": "7.4"}}` from `renovate.json` (I see it in "migrated config" section of DEBUG output)
- packages still are updated on PHP 8.2 runtime and bumped to ones not compatible with 7.4

The reason why I want to do that is that we're migrating our app, we have 8.2 on production but some of machines still needs migration (code is compatible with both 7.4 and 8.2 thanks to PHPCompatibility sniffs). We want to use 8.2 for local development (only as runtime, without using features at this point), but disallow Renovate from bumping packages to versions not compatible with 7.4.

@rarkins any ideas?

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/74ad0338964e19b18be6267b5b8e618f7ffbade4/lib/modules/manager/composer/artifacts.ts#L133-L136

it should clearly prefer the constraints from config. can you please provide a minimal reproduction?

**Reply by viceice** [reply]
we recently fixed a bug where extracted constraints had overwritten the configured constraints. so please try latest renovate version.

**Reply by Wirone** [reply]
I'll try to do it later today or tomorrow.

**Reply by Wirone** [reply]
@viceice I reproduced the problem in a repo:
- [run without forced constraints](https://gitlab.com/codito-dev/issue-reproducers/renovate-bot/constraints-override/-/jobs/4589855492#L229): Renovate tries to update `ramsey/uuid` to 4.7.4 and that's OK because platform is 8.2.7
- [run with forced constraint](https://gitlab.com/codito-dev/issue-reproducers/renovate-bot/constraints-override/-/jobs/4589893601#L229): still trying to update that package even though `force.constraints.php` was set

PS. In [this run](https://gitlab.com/codito-dev/issue-reproducers/renovate-bot/constraints-override/-/jobs/4589549530) you can see package wasn't updated even without `force.constraints.php`, but there wasn't `packageRules` defined, I've added them later (based on our real setup) to reproduce updating flow. I think especially [this](https://gitlab.com/codito-dev/issue-reproducers/renovate-bot/constraints-override/-/blob/d356e631b6ac543a78a519cdf4eab9501ffed184/renovate.json#L25-L31) may be related.

Maybe I'm doing something wrong, but I would expect packages are updated only within forced PHP constraint.

**Reply by rarkins** [reply]
force.constraints should not be required. constraints should itself always take precedence (and we fix if not)

**Reply by Wirone** [reply]
@rarkins I've tried without `force` too, but it does not work anyway. Renovate considers 8.2 as platform (or does not take platform into consideration, I don't know) and tries to bump package that shouldn't be bumped.

**Reply by rarkins** [reply]
I don't think this is related to the original topic of this discussion. You seem to be referring to `constraintsFiltering` and if so then please start a new discussion

**Comment by rarkins** [comment]
Reproduction forked, tested here: https://github.com/renovate-reproductions/21204/pull/2

From logs we can see that constraint `^8.1` is used (presumably from `require.php`).

```
DEBUG: packageFiles with updates (repository=renovate-reproductions/21204, baseBranch=main)
       "config": {
         "composer": [
           {
             "constraints": {"php": "^8.1"},
```

Fix: https://github.com/renovatebot/renovate/pull/21207
